Technotes/001-async-networking.md

```diff
@@ -0,0 +1,246 @@
+
+The `ImageFetcherTask` provides access to the `configuration`, `handler`, and `result` of a given operation. While it seems like it should be very unlikely, this present a potential thread-safety issue as it is possible for the caller to try reading a task’s `result` on one thread while the underlying `Operation`’s `completionBlock` mutates the task’s `result` on another thread or for the caller to mutate its `configuration` or `handler` member while it is being read on another thread.
+
+#### Proposed `tasks` Implementation
+
+This proposal changes the `tasks` collection to a `Dictionary` and first checks for an active/pending for the requested configuration before checking for a cached image or requesting one. This absolves the caller of responsibility for preventing duplicate requests, but it also means that any call to cancel an active / pending request cancels it for all callers.
+
+I replaced `ImageFetcherTask` by providing direct access to the underlying `Task`, which offers access to the request’s eventual value and a `cancel()` method. The result is more restricted but safer since we remove possibility of concurrently reading/writing task members.
+
+`ImageFetcher` will use an `NSLock` to ensure thread-safe access to `tasks`. I considered making it an `actor` but that seems like overkill when we only need to protect `tasks` and all access to it will be very quick.
+
+### Managing Concurrency
+
+This proposal also removes the ability to limit concurrent requests and cache lookups via `maxConcurrent` by making the cooperative thread pool responsible for managing those operations. Instead, `maxConcurrent` is limited to controlling concurrent processing of image data ultimately performed by the `Image.decompress(...)` method. I reasoned that this represents the most expensive operation performed by `ImageFetcher` and the one the OS is least suited to manage. [Previous efforts](https://github.com/Mobelux/ImageFetcher/pull/6#issuecomment-1115291685) at limiting concurrency outside of an `OpertionQueue` were not especially fruitful.
+
+## Performance
+
+The proposed design only tries to limit concurrency when processing images. I used both the current and the proposed library to load images from JSONPlaceholder in a collection view:
+
+![Demo App View](images/001-profiled-app-view.jpeg)
+
+I profiled this on an iPad Pro with `maxConcurrent = 2` for both libraries to compare performance and resource usage.
+
+### Current Implementation
+
+CPU Usage (via Time Profiler) and Requests:
+
+![Current: Time Profiler](images/001-time-profiler-current.png)
+
+Zoomed out to fit all requests:
+
+![Current: Zoomed Out](images/001-time-profiler-current-zoom-out.png)
+
+Activity per CPU:
+
+![Current: CPU](images/001-activity-current.png)
+
+### Proposed Implementation
+
+CPU Usage (via Time Profiler) and Requests:
+
+![Proposed: Time Profiler](images/001-time-profiler-proposed.png)
```

Comment:
I interpret this as being many concurrent threads, which I don't fully understand since the cooperative thread pool is bound to the number of "available" cpu cores.  What am I missing?

Reply:
If you are referring to all the green bars, those just show the lifetime of tasks, not their thread usage. I should have included a detail view of those tasks:

![Detail](https://user-images.githubusercontent.com/422815/231257720-b492f8a8-8392-4fc5-a048-19a97cbddb97.png)